File: SDE_sheet/Arrays/08_merge_intervals/solution.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: x[0])
        res = []
        
        i = 0
        while i < len(intervals):
            start, end = intervals[i]
            
            while i + 1 < len(intervals) and end >= intervals[i + 1][0]:
                i += 1
                end = max(end, intervals[i][1])
            
            res.append([start, end])
            i += 1
        
        return res

File: SDE_sheet/Arrays/08_merge_intervals/solution.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        List<int[]> res = new ArrayList<>();
        
        for (int i = 0; i < intervals.length; i++) {
            int start = intervals[i][0];
            int end = intervals[i][1];
            
            while (i + 1 < intervals.length && end >= intervals[i + 1][0]) {
                i++;
                end = Math.max(end, intervals[i][1]);
            }
            
            res.add(new int[]{start, end});
        }
        
        return res.toArray(new int[res.size()][]);
    }
}
